core/src/main/java/io/grpc/internal/GzipInflatingBuffer.java

```diff
@@ -0,0 +1,514 @@
+   * unable to produce more output.
+   */
+  boolean isStalled() {
+    checkState(!closed, "GzipInflatingBuffer is closed");
+    return gzippedData.readableBytes() == 0 && (inflaterInputStart == inflaterInputEnd);
+  }
+
+  /**
+   * Returns true when there is gzippedData that has not been input to the inflater or the inflater
+   * has not consumed all of its input.
+   */
+  boolean hasPartialData() {
+    checkState(!closed, "GzipInflatingBuffer is closed");
+    return gzippedData.readableBytes() != 0 || (inflaterInputStart < inflaterInputEnd);
+  }
+
+  /**
+   * Adds more gzipped data, which will be consumed only when needed to fulfill requests made via
+   * {@link #inflateBytes}.
+   */
+  void addGzippedBytes(ReadableBuffer buffer) {
+    checkState(!closed, "GzipInflatingBuffer is closed");
+    gzippedData.addBuffer(buffer);
+  }
+
+  @Override
+  public void close() {
+    if (!closed) {
+      closed = true;
+      gzippedData.close();
+      if (inflater != null) {
+        inflater.end();
+        inflater = null;
+      }
+    }
+  }
+
+  /**
+   * Attempts to inflate up to {@code n} bytes of data into {@code bufferToWrite}. This method will
+   * always write as many inflated bytes as it can produce, up to a maximum of {@code
+   * bytesRequested}.
+   *
+   * <p>This method may consume gzipped bytes without writing any data to {@code bufferToWrite}, and
+   * may also write data to {@code bufferToWrite} without consuming additional gzipped bytes (if the
+   * inflater on an earlier call consumed the bytes necessary to produce output).
+   *
+   * @param n max number of bytes to inflate
+   * @param bufferToWrite destination for inflated data
+   * @return gzipped bytes consumed by the call (NOT the number of inflated bytes written)
+   */
+  int inflateBytes(int n, CompositeReadableBuffer bufferToWrite)
+      throws DataFormatException, ZipException {
+    checkState(!closed, "GzipInflatingBuffer is closed");
+
+    int bytesNeeded = n;
+    while (bytesNeeded > 0) {
+      int bytesWritten = fillInflatedBuf(bytesNeeded);
+      if (bytesWritten == 0) {
+        break;
+      } else {
+        bytesNeeded -= bytesWritten;
+        writeInflatedBufToOutputBuffer(bufferToWrite);
+      }
+    }
+    int savedBytesConsumed = bytesConsumed;
+    bytesConsumed = 0;
+    return savedBytesConsumed;
+  }
+
+  private int fillInflatedBuf(int n) throws DataFormatException, ZipException {
+    int bytesToInflate;
+    if (inflaterOutput == null) {
+      bytesToInflate = Math.min(n, MAX_OUTPUT_BUFFER_SIZE);
+      inflaterOutput = new byte[bytesToInflate];
+    } else {
+      // Empty buffer allocated on previous call
+      bytesToInflate = Math.min(n, inflaterOutput.length);
+    }
+
+    int bytesNeeded;
+    boolean madeProgress = true;
+    while (madeProgress && (bytesNeeded = bytesToInflate - inflaterOutputEnd) > 0) {
+      switch (state) {
+        case HEADER_MAGIC:
```

Comment:
That's a _lot_ of cases for the HEADER. Can we reduce it by a few?

Reply:
Condensed the three cases for header magic, compression method, and remaining bytes. The rest of the cases are cleaner to keep separated, I think